Initial commit for MetricsSink operators.

nelsonong opened the pull request.

m-kotowski commented on README.md:
```diff
@@ -58,3 +58,22 @@ command:
+3.  Run `ant maven-deps` to download the dependencies required by the project
+3.  In Streams Studio, Import...
+5.  In the dialog, select InfoSphere Streams -> SPL Project
+6.  Select com.ibm.streamsx.mail to import the project into studio.
```
Do you mean com.ibm.streamsx.mail or com.ibm.streamsx.metrics?

nelsonong replied:
Yes, I mean com.ibm.streamsx.metrics. I have read all your comments and will update and create a new pull request soon with fixes to the problems you've found.

m-kotowski commented on com.ibm.streamsx.metrics/build.xml:
```diff
@@ -0,0 +1,99 @@
+
+	<!-- Downloads libraries using maven -->
+	<target name="maven-deps" depends="init,nomaven" if="env.M2_HOME">
+		<fail unless="pomset" message="There must be a pom.xml file in com.ibm.streamsx.hbase in order to use maven to download necessary dependencies.  Three sample pom files are included: pom-v094.xml (hbase 0.94, hadoop 1),pom-v096-hadoop1.xml (hbase 0.96, hadoop 1), and  pom-v096-hadoop2.xml (hbase 0.96, hadoop 2).  Please rename the correct one for your install to pom.xml"/> 
```
Do you need streamsx.hbase? I assume that you copied this file from another repo that also copied it from another one. Please cleanup this file, so also messages and comments are correct.

m-kotowski commented on com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/CustomMetricsSink.java:
```diff
@@ -0,0 +1,168 @@
+import com.ibm.streams.operator.model.PrimitiveOperator;
+
+@PrimitiveOperator(name="CustomMetricsSink", namespace="com.ibm.streamsx.metrics",
+description="Java Operator CustomMetricsSink")
```
Please provide a more detailed operator description as well as a more detailed port description.

ddebrunner replied:
Not sure about the name `CustomMetricSink`, but not really sure what it does. How does a user customize it? Is it really something like a Metrics2Something?

nelsonong replied:
Currently, the MetricsSink outputs the metrics to a database that Grafana and Kibana can query from.
Pete suggested that it might be useful for users to output their own custom data (not necessarily metrics) to the database. For example, I connected it to a health application which outputted heart rate and ECG data to the database.
So, I quickly made this operator which simply takes tuple attribute names as a parameter and outputs it to the database. Not sure what the name of the operator should be or whether we should keep it in the metrics toolkit.

ddebrunner replied:
> simply takes tuple attribute names as a parameter and outputs it to the database

Seems like it has nothing to do with metrics, in the way Streams uses the term 'metric'.

Is it just a Sink for Grafana and Kibana?

Once there's a full description of what it does it will be easier to name it, and determine if it belongs in this toolkit.

ddebrunner replied:
Just to add "custom metric" has a specific meaning in IBM Streams to mean metrics defined by an operator. We should not use the same term to mean something else.

nelsonong replied:
Yes, I agree, it's just a sink for Grafana/Kibana and doesn't have to do with metrics. I'll take it out.

m-kotowski commented on com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/CustomMetricsSink.java:
```diff
@@ -0,0 +1,168 @@
+	// http://www.ibm.com/support/knowledgecenter/en/SSCRJU_4.2.0/com.ibm.streams.dev.doc/doc/jmxapi-start.html
+	// Environment variables (@...@) are evaluated during compile-time and must
+	// be identical in the run-time environment.
+	"@STREAMS_INSTALL@/lib/com.ibm.streams.management.jmxmp.jar",
```
Do you also use JMX? If not, cleanup this section.

m-kotowski commented on com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/CustomMetricsSink.java:
```diff
@@ -0,0 +1,168 @@
+        Logger.getLogger(this.getClass()).trace("Operator " + context.getName() + " initializing in PE: " + context.getPE().getPEId() + " in Job: " + context.getPE().getJobId() );
+	
+        // Connect to InfluxDB server.
+        influxDB = InfluxDBFactory.connect("http://localhost:8086", "admin", "admin");
```
Streams is running on several servers. So, what if your influx DB is not running on the local host. You cannot and must not assume that the user/password is admin/admin.

m-kotowski commented on com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/MemoryMetricsSink.java:
```diff
@@ -0,0 +1,333 @@
+	// http://www.ibm.com/support/knowledgecenter/en/SSCRJU_4.2.0/com.ibm.streams.dev.doc/doc/jmxapi-start.html
+	// Environment variables (@...@) are evaluated during compile-time and must
+	// be identical in the run-time environment.
+	"@STREAMS_INSTALL@/lib/com.ibm.streams.management.jmxmp.jar",
```
Cleanup required if you do not use JMX.

m-kotowski commented on com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/MemoryMetricsSink.java:
```diff
@@ -0,0 +1,333 @@
+        Logger.getLogger(this.getClass()).trace("Operator " + context.getName() + " initializing in PE: " + context.getPE().getPEId() + " in Job: " + context.getPE().getJobId() );
+        
+        // Connect to InfluxDB server.
+        influxDB = InfluxDBFactory.connect("http://localhost:8086", "admin", "admin");
```
URL, user, and password must not be hardcoded.

m-kotowski commented on com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/MemoryMetricsSink.java:
```diff
@@ -0,0 +1,333 @@
+            .build();
+    	
+    	// Connect to ElasticSearch server.
+		client = new PreBuiltTransportClient(Settings.EMPTY).addTransportAddress(new InetSocketTransportAddress(InetAddress.getByName("localhost"), 9300));
```
URL and port are not fixed, are they?

m-kotowski commented on com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/MemoryMetricsSource.java:
```diff
@@ -0,0 +1,535 @@
+	// ------------------------------------------------------------------------
+	
+	static final String DESC_OPERATOR = 
+			"The MemoryMetricsSource operator uses the "
```
The intention of the MetricsSource operator was to provide all kind of metrics, also the PE metrics. Therefore, please enhance the MetricsSource operator with the PE metrics instead of creating a new operator.

m-kotowski commented on com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/MetricsSink.java:
```diff
@@ -0,0 +1,157 @@
+import org.influxdb.dto.Point;
+
+@PrimitiveOperator(name="MetricsSink", namespace="com.ibm.streamsx.metrics",
+description="Java Operator MetricsSink")
```
detailed description ...

m-kotowski commented on com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/MetricsSink.java:
```diff
@@ -0,0 +1,157 @@
+
+@PrimitiveOperator(name="MetricsSink", namespace="com.ibm.streamsx.metrics",
+description="Java Operator MetricsSink")
+@InputPorts({@InputPortSet(description="Port that ingests tuples", cardinality=1, optional=false, windowingMode=WindowMode.NonWindowed, windowPunctuationInputMode=WindowPunctuationInputMode.Oblivious), @InputPortSet(description="Optional input ports", optional=true, windowingMode=WindowMode.NonWindowed, windowPunctuationInputMode=WindowPunctuationInputMode.Oblivious)})
```
detailed description ...

m-kotowski commented on com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/MetricsSink.java:
```diff
@@ -0,0 +1,157 @@
+	// http://www.ibm.com/support/knowledgecenter/en/SSCRJU_4.2.0/com.ibm.streams.dev.doc/doc/jmxapi-start.html
+	// Environment variables (@...@) are evaluated during compile-time and must
+	// be identical in the run-time environment.
+	"@STREAMS_INSTALL@/lib/com.ibm.streams.management.jmxmp.jar",
```
Cleanup if not needed ...

m-kotowski commented on samples/com.ibm.streamsx.metrics.sample.memorymonitor/application/Main.spl:
```diff
@@ -0,0 +1,44 @@
+				hostname : "smtp.gmail.com" ;
+				subject : "Memory leak detected!" ;
+				to : "2263748929@txt.bell.ca" ;
+				username : "nelsonong3@gmail.com" ;
```
Do you really want to put your email account and password in the open???

ddebrunner commented on com.ibm.streamsx.metrics/build.xml:
```diff
@@ -0,0 +1,99 @@
+<!--
+# Licensed Materials - Property of IBM
+# Copyright IBM Corp. 2011, 2013
+# US Government Users Restricted Rights - Use, duplication or
```
This is actually the wrong copyright for open source projects, remove the US Government restrictions since the code is released under ALv2.

nelsonong replied:
Okay, I'll do that.

m-kotowski commented on com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/MemoryMetricsSource.java:
```diff
@@ -0,0 +1,535 @@
+ * <p>With the exception of operator initialization, all the other events may occur concurrently with each other, 
+ * which lead to these methods being called concurrently by different threads.</p> 
+ */
+@PrimitiveOperator(
```
This operator becomes obsolete with the latest (already merged) MetricsSource extension. The MetricsSource operator now supports also PE metrics.

m-kotowski commented on com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/internal/TupleContainer.java:
```diff
@@ -227,6 +232,17 @@ public void setJobName(String jobName) {
+	 * 
+	 * @param peId
+	 */
+	public void setPeId(BigInteger peId) {
```
You shouldn't be able to compile because the same method exists in line 292 in the master.

m-kotowski commented on com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/internal/TupleContainer.java:
```diff
@@ -40,7 +40,7 @@
 
 	/**
-	 * Index of the resource attribute.
+   * Index of the resource attribute.
```
Please revert to the original indent.

m-kotowski commented on com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/internal/TupleContainer.java:
```diff
@@ -134,6 +134,11 @@ public TupleContainer(StreamingOutput<OutputTuple> port) {
 			_jobNameAttributeIndex = Integer.valueOf(attribute != null && attribute.getType().getMetaType() == Type.MetaType.RSTRING ? attribute.getIndex() : -1) ;
 		}
+		// PE-related attributes.
+		if (_peIdAttributeIndex  == null) {
```
This code is already present in line 151+ in the master.

m-kotowski commented on com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/MetricsSink.java:
```diff
@@ -0,0 +1,362 @@
+@Libraries({
+	"opt/downloaded/*"
+	})
+public class MetricsSink extends AbstractOperator {
```
Is MetricsSink a good name? On the one hand, a [MetricsSink](http://www.ibm.com/support/knowledgecenter/SSCRJU_4.2.0/com.ibm.streams.toolkits.doc/spldoc/dita/tk$spl/op$spl.adapter$MetricsSink.html) operator already exists. On the other hand, you write data to one or two databases.

It seems as if sending to InfluxDB is optional, why? And why is ElasticSearch mandatory?

I wonder whether it makes sense to (a) split the operator - ElasticSearchSink for [ElasticSearch](https://www.elastic.co/products/elasticsearch) and InfluxDBSink for [InfluxDB](https://www.influxdata.com/time-series-platform/influxdb/), and (b) to follow a more generic approach, which means: evaluate the input stream's meta data and send the complete tuple to the underlying database.

From what I can see in the code, the key is the attribute name except for the metricName-metricValue pair, and the value is always the attribute value except for lastTimeRetrieved, which is handled as timestamp. Is there a generic approach? For example, introduce a parameter to name the timestamp attribute. For the metricName-metricValue pair I have no idea for a generic approach so far, but perhaps you have one!?

If we want to stay with metrics-specific sinks, perhaps MetricsElasticSearchSink and MetricsInfluxDBSink might be useful names.

ddebrunner replied:
There's no requirement for a sink operator to have 'Sink' in the name. The name should reflect the action(s) of the operator, in this case typically in the terminology of the system being written to. E.g. the sink operator that pushes tuples to Kafka is a `KafkaProducer` as "producer" is the terminology use by Apache Kafka.

nelsonong replied:
Yes, I talked with Pete and he said it was his intention all along to split the operators. I'll stick with "MetricsElasticSearchSink" and "MetricsInfluxDBSink" for now and can change it later on if we decide on better names.

And yes, I can make a more generic approach for outputting the complete tuple to the underlying databases. For the metricName-metricValue pair, I'll just output the metrics as-is without joining them. This just means that querying for metric values in the databases would go from looking like this:
`SELECT "<metricName>" FROM "<databaseName>"`

to something more like this:
`SELECT "metricValue" FROM "<databaseName>" WHERE "metricName"="<metricName>"`

m-kotowski commented on README.md:
````diff
@@ -58,3 +58,22 @@ command:
 ```
+
+### Building Project
+The MetricsSink operator requires you to build jars and include them in your project. You will also need to have an ElasticSearch server running.
````
Can you provide some links that help to setup the ElasticSearch server? And what is with the InfluxDB?

m-kotowski replied:
Please provide at least one sample application for your operator. For this sample, please also describe the required preparation, so that somebody can import and run the SPL sample.

ddebrunner replied:
Why is it required that the user has to build jars, why can't a released toolkit already contain the jars?

nelsonong replied:
I followed the example of the [streamsx.mail](https://github.com/IBMStreams/streamsx.mail) toolkit. It might be to make the toolkit more compact. I can include the jars in my next push.

ddebrunner replied:
To be clear, if the jars are built from the java files in the toolkit they should not be added to the source tree. Instead a release is produced which is a built version of the toolkit.

Ie. we should make the toolkit optimal for users that download a pre-built release of it.

I'm not sure what jars are being talked about.

nelsonong replied:
The jars that are built are for the ElasticSearch and InfluxDB Java API's needed for the MetricsSink operators to communicate and add data to the databases. There's quite a few jars (~30 MB, ~50 jars) which is why I thought excluding them would be for the best.

If I include them, it would definitely make it easier on the users, though.

m-kotowski commented on com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/internal/TupleContainer.java:
```diff
@@ -40,7 +40,7 @@
 	private Integer _jobNameAttributeIndex = null;
```
It seems to me as if this file can be taken from master as-is. See my other comments for this file.

ddebrunner commented:
I do wonder if the ElasticSink & InfluxDB operators should be moved into separate toolkits as generic operators to insert into those databases. Then there can be operators in this toolkit that convert metrics into the data format expected by those operators.

Otherwise every application will have to pay the overhead of the elastic and influxdb jars in their sab files, even if they are not using them.

nelsonong commented:
Yes, there's a decent amount of overhead so this might be a good idea. Would this new toolkit remain in this repository (seeing as it still depends on metrics and the MetricsSource operator)?

m-kotowski commented:
re moving the Elastic and InfluxDB operators:

There is a discussion ongoing whether toolkits shall be kept small (for example, a separate toolkit per each 3rd party product), or whether similar things shall go into the same toolkit (for example, ElasticSearch and InfluxDB are both databases). See also [here](https://github.com/IBMStreams/administration/issues/101).

I expect that every 3rd party product will go into its own, separate toolkit. Therefore, **if both operators can be made generic enough**, they could go into either their own toolkits (ElasticSearch toolkit & InfluxDB toolkit), which have to be requested via an administration issue [here](https://github.com/IBMStreams/administration/issues).

But, if they cannot made generic enough and if they are very specific to metrics, they should stay in this toolkit. I wonder what @mikespicer 's opinion is on this topic because he requested to add metrics sink operators here, see issue #10.

ddebrunner commented:
There would still be metrics sink operators in this toolkit, they would wrap the generic operators in a composite, their role would be to convert the metrics into the desired database format.

mikespicer commented:
I agree with @ddebrunner 
Experience from the messaging toolkit has moved us more to separating out technologies like Kafka and RabbitMQ to improve isolation with different update schedules for different technologies. Having them separate increases dependencies and the number of toolkits you need to download but we could solve that by making it easier to determine dependencies and download all toolkits needed as a group.

nelsonong commented:
I have made the ElasticSearch and InfluxDB sinks generic. They are generic enough to be attached to the MetricsSource without any wrapper. Should I still create composite wrappers for these?

Also, if the metrics sink operators will go into this toolkit, where will the generic sinks go? Into their own toolkits that will act as a dependency for this toolkit?

ddebrunner commented:
I don't know enough about ElasticSearch and InfluxDB but are the sink operators really generic or specific to metrics still, since they have 'Metrics' in their names. Are these databases only intended for metrics (seems unlikely).

If the operators are truly generic then they would handle an arbitrary scheme (maybe within some restrictions, such as having a timestamp) that might not be metrics, e.g. just a generic timeseries. 

Metrics composites would call them setting the operator parameters to save the relevant data from the metrics tuple schema.

nelsonong commented:
Yes, they simply read the incoming tuple's attributes and values and outputs them to the databases. All it needs is the server info (hostname, hostport) and database info (database name, table name, timestamp name).

It should work with any app, not just metrics-specific ones and it doesn't need any special configuration for metrics.

ddebrunner commented:
Since these are Java operators why are the generated operator models XML being added to source control?

ddebrunner commented:
> Yes, they simply read the incoming tuple's attributes and values and outputs them to the databases.

But that's not what there descriptions say:

```
The MetricsElasticSearchDBSink operator receives metrics as tuples from "
 +			+ "the MetricsSource operator and outputs these metrics to an ElasticSearch "
```

nelsonong commented:
Yes, sorry, that's old code. I have, since, changed the operator names to ElasticSearchSink and InfluxDBSink and changed their descriptions. I just wanted to confirm whether a wrapper should be added and whether the generic operators should go in their own toolkit before pushing.

And yes, sorry. I will remove the operator model XMLs.

m-kotowski commented:
@nelsonong I recommend following procedure:

Go to https://github.com/IBMStreams/administration/issues

Raise two issues: 
1. Proposal: Elasticsearch toolkit
2. Proposal: InfluxDB toolkit

Provide for both toolkit proposals a summary why you request them and what the initial content will be. You can also mention @mikespicer 's and @ddebrunner 's comments. You should mention that they are generic and not related to, for example, the metrics use case only.

Then, as soon as the toolkits are created, you move your Sink operators to the corresponding toolkit. I recommend proper namespaces and operator names like com.ibm.streamsx.elasticsearch::ElasticsearchSink and com.ibm.streamsx.influxdb::InfluxDBSink

As Dan mentioned, if Elasticsearch or InfluxDB have special names for their "data providers", you should use these terms instead of Sink.

Finally, provide samples, SPLDOC documentation, installation support (linking to the homepages) etc. and you are done.

If these operators can be used as-is also for our "metrics use case", then there is no need to implement wrappers. But, you could add some SPLDOC documentation sketching how to connect MetricsSource and your sink operators, for examples, some samples - one for elasticsearch, one for influxdb

At the end, please cancel pull request #22, or can we cancel it immediately?

rnostream commented:
In Bluemix there is already a metric and log collection/visualization service available. This is Grafana and Kibana based but has own API for data provisioning. This would need separate MetricSink.

markheger commented:
Close because proposed Sinks should be stored in separate repositories.
